dev/buildtool/errors.py

```diff
@@ -0,0 +1,105 @@
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Helper functions for raising and logging errors."""
+
+
+import logging
+import os
+import traceback
+
+from buildtool.metrics import MetricsManager
+
+
+class BuildtoolError(Exception):
+  RUNTIME = 'runtime'
+  USAGE = 'usage'
+  INTERNAL = 'internal'
+
+  def __init__(self, message, classification, cause=None):
+    super(BuildtoolError, self).__init__(message)
+    labels = {
+        'cause': cause,
+        'classification': classification
+    }
+    MetricsManager.singleton().inc_counter(
+        'BuildtoolError', labels, 'Errors raised during execution.')
+
+
+class ConfigError(BuildtoolError):
+  def __init__(self, message, cause=None):
+    super(ConfigError, self).__init__(
+        message, self.INTERNAL, cause=cause or 'config')
+
+class TimeoutError(BuildtoolError):
+  def __init__(self, message):
+    super(TimeoutError, self).__init__(message, self.RUNTIME)
+
+class ExecutionError(BuildtoolError):
+  def __init__(self, message, program=None):
+    super(ExecutionError, self).__init__(message, self.RUNTIME, cause=program)
+
+class ResponseError(BuildtoolError):
+  def __init__(self, message, server=None):
+    super(ResponseError, self).__init__(message, self.RUNTIME, cause=server)
+
+class UnexpectedError(BuildtoolError):
+  def __init__(self, message, cause=None):
+    super(UnexpectedError, self).__init__(message, self.INTERNAL,
+                                          cause=cause or 'unexpected')
+
+
+def maybe_log_exception(where, ex, action_msg='propagating exception'):
+  """Log the exception and stackdrace if it hasnt been logged already."""
+  if not hasattr(ex, 'loggedit'):
+    text = traceback.format_exc()
+    logging.error('"%s" caught exception\n%s', where, text)
+    ex.loggedit = True
+
+  logging.error('"%s" %s', where, action_msg)
```

Comment:
Does `logging` not support the bracket notation for string formatting? It's kinda funny - java does it the exact opposite way, regular string formatting requires the `%s`, but `java.logging` supports brackets...

Reply:
python brackets in strings are resolved by ".format" method on the string. logging takes a string parameter, so I dont think you can use brackets. you could use brackets to pass the string parameter in so logging sees a single string arg. Style guides discourage that because it forces strings to be resovlved eagerly rather than lazily so you cant avoid it if you disable logging. That's not a concern here, but feels simpler to do that way.

Reply:
Makes sense just curious.